### PR TITLE
Fix deletes which trigger file rewrites in partitions with special characters

### PR DESCRIPTION
**Pull request by jkylling**
## Description

Fix deletes which trigger file rewrites in partitions with special characters.

## Additional context and related issues

Fixes https://github.com/trinodb/trino/issues/18922

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix deletes in Delta tables with partitions with special characters. ({issue}`18922`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
> Please see https://github.com/trinodb/trino/issues/18922

Please change to `Fixes #18922` so that we can close the issue automatically.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 1]
```diff
@@ -1919,7 +1919,7 @@ public void finishMerge(ConnectorSession session, ConnectorMergeTableHandle merg
             }
 
             for (String file : oldFiles) {
-                transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(file, writeTimestamp, true));
+                transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(toUriFormat(file), writeTimestamp, true));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1922]: There's another method `executeDelete` doing DELETE statement. Please add a new test case.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 2]
```diff
@@ -2108,7 +2108,7 @@ private void finishOptimize(ConnectorSession session, DeltaLakeTableExecuteHandl
 
             for (String scannedPath : scannedPaths) {
                 String relativePath = relativePath(tableLocation, scannedPath);
-                transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(relativePath, writeTimestamp, false));
+                transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(toUriFormat(relativePath), writeTimestamp, false));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2111]: Which test cover this change?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 1 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1922]: I've extended the current test to also do a non-targeted `DELETE FROM`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 2 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2111]: I've extended the existing `testOptimize` test to have rows with partition values `with:colon`.

**Comment by ebyhr** [comment]
/test-with-secrets sha=5abfe4968881af2dcded05658a74094cbac0feec

**Comment by jkylling** [comment]
@ebyhr We get one failure on
```
TestDeltaLakeAdlsConnectorSmokeTest>BaseDeltaLakeConnectorSmokeTest.testOptimizeWithPartitionedTable:1790->AbstractTestQueryFramework.computeActual:299->AbstractTestQueryFramework.computeActual:304 » QueryFailed Error opening Hive split abfs://***@***.dfs.core.windows.net/test-delta-lake-integration-smoke-test-ldwyxhzdn6/test_optimize_partitioned_table_aa992pntr1/value=with:colon/20230907_065206_00472_8ccir_899be84e-0922-4983-b5a1-e2a381f0fbad (offset=0, length=201): abfs://***@***.dfs.core.windows.net/test-delta-lake-integration-smoke-test-ldwyxhzdn6/test_optimize_partitioned_table_aa992pntr1/value=with:colon/20230907_065206_00472_8ccir_899be84e-0922-4983-b5a1-e2a381f0fbad
```
It looks similar to https://github.com/trinodb/trino/issues/15376, https://github.com/trinodb/trino/pull/15381

I believe the failure is just because we use `:` and not `+` or space in the path, and we have no special logic to handle `:`. We could change the test to use `+`, or omit the commit which changes the optimize test?

**Comment by jkylling** [comment]
@ebyhr I've removed the commit adding a test for the OPTIMIZE procedure, as its failure is a separate issue. Would you have time for another review and another run of the tests with secrets?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java [hunk 3]
```diff
@@ -67,9 +67,12 @@ public void testTargetedDeleteWhenTableIsPartitionedWithColumnContainingSpecialC
                 "AS VALUES " +
                 "(1, 'with-hyphen'), " +
                 "(2, 'with:colon'), " +
-                "(3, 'with?question')", 3);
+                "(3, 'with:colon'), " +
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java:70]: why `with:colon` twice?

Comment (jkylling) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java:70]: By having `with:colon` twice we get a parquet file with two rows. When we next delete one of these rows we trigger a "rewrite" of the parquet file, and corresponding remove and add entries in the delta log. If we only have a single row in the parquet file we would not need to do a rewrite of the parquet file and just add a remove entry in the delta log. Previously the code path which handles rewrites of parquet files did not have test coverage for special characters.

Comment (findepi) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java:70]: Capture please the gist of this a code comment

Comment (jkylling) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java]: Added a comment and an assertion that the rows are in a single parquet file.

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java [hunk 4]
```diff
@@ -67,9 +67,13 @@ public void testTargetedDeleteWhenTableIsPartitionedWithColumnContainingSpecialC
                 "AS VALUES " +
                 "(1, 'with-hyphen'), " +
                 "(2, 'with:colon'), " +
-                "(3, 'with?question')", 3);
+                "(3, 'with:colon'), " + // create two rows in a single file to trigger parquet file rewrite on delete
+                "(4, 'with?question')", 4);
+        assertQuery("SELECT count(*), count(DISTINCT \"$path\"), col_name FROM " + tableName + " GROUP BY 3", "VALUES(1, 1, 'with-hyphen'), (2, 1, 'with:colon'), (1, 1, 'with?question')");
         assertUpdate("DELETE FROM " + tableName + " WHERE id = 2", 1);
-        assertQuery("SELECT * FROM " + tableName, "VALUES(1, 'with-hyphen'), (3, 'with?question')");
+        assertQuery("SELECT * FROM " + tableName, "VALUES(1, 'with-hyphen'), (3, 'with:colon'), (4, 'with?question')");
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java:74]:
```suggestion
assertQuery("SELECT * FROM " + tableName, "VALUES (1, 'with-hyphen'), (3, 'with:colon'), (4, 'with?question')");
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java [hunk 5]
```diff
@@ -67,9 +67,13 @@ public void testTargetedDeleteWhenTableIsPartitionedWithColumnContainingSpecialC
                 "AS VALUES " +
                 "(1, 'with-hyphen'), " +
                 "(2, 'with:colon'), " +
-                "(3, 'with?question')", 3);
+                "(3, 'with:colon'), " + // create two rows in a single file to trigger parquet file rewrite on delete
+                "(4, 'with?question')", 4);
+        assertQuery("SELECT count(*), count(DISTINCT \"$path\"), col_name FROM " + tableName + " GROUP BY 3", "VALUES(1, 1, 'with-hyphen'), (2, 1, 'with:colon'), (1, 1, 'with?question')");
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeDelete.java:72]:
```suggestion
assertQuery("SELECT count(*), count(DISTINCT \"$path\"), col_name FROM " + tableName + " GROUP BY 3", "VALUES (1, 1, 'with-hyphen'), (2, 1, 'with:colon'), (1, 1, 'with?question')");
```

**Review by findinpath — APPROVED** [review]

**Comment by findepi** [comment]
/test-with-secrets sha=933ea36acb7ba9df0d49e27e50575ef33f6584d6

**Comment by ebyhr** [comment]
/test-with-secrets sha=933ea36acb7ba9df0d49e27e50575ef33f6584d6

**Review by ebyhr — APPROVED** [review]

**Pull request merged**